[Wen Xuan] iP

e1121208 opened the pull request:
# Axel

> “You don't have to control your thoughts. You just have to stop letting them control you.” – Dan Millman

**Axel** frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- ~~fast~~ ~~SUPER FAST~~ **INSANELY FAST** to use

All you need to do is:

1. Download it from [here](https://github.com/e1121208/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you! :wink:

And it's **FREE**!

## Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too! Here's an example of the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Renamed Duke to Axel

Commit:
Implemented a greeting and exit message

Commit:
Added echo feature

Commit:
Added the feature to add texts to a limited list and display them

Commit:
Created a new Task class

Commit:
Implemented the feature for marking and unmarking of whether tasks in the list are done

Commit:
Created child classes under the parent 'Task' class

Commit:
Added the feature to track to do, deadline and event tasks

Commit:
Created two specific exception classes

Commit:
Added an exception for unrecognised commands

Commit:
Further tidied the exception classes, and implemented features to handle exceptions. Added a little character to Axel too. Debugged the toDo function.

Commit:
Removed redundant MarkException and rephrased to InvalidTaskNumberException for better usage and clarity, added the feature of deletion and the use of list instead of an array

Commit:
Add testing of text UIs

Commit:
Add the features of storing the data of previous runs and reloading that data on the next run when Axel is booted again

Commit:
Merge branch 'branch-Level-7'

Commit:
Update Axel to understand dates and times

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/EventTask.java

Commit:
Use the input/output redirection technique to semi-automate the testing of Duke

Commit:
Extract Ui, Storage, Parser, TaskList and Command classes to make the code more OOP

Commit:
Move classes into a package

Commit:
Merge branch 'master' into add-gradle-support

Commit:
Automate project builds using Gradle

Commit:
Add JUnit tests

Commit:
Debug the code further to package the app as a JAR file

Commit:
Add JavaDoc comments

Commit:
Tweak the code to comply with a coding standard

Commit:
Give users a way to find a task by searching for a keyword in the task description

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/axel/Axel.java
#	src/main/java/axel/Command.java
#	src/main/java/axel/Parser.java
#	src/main/java/axel/Storage.java
#	src/main/java/axel/Task.java
#	src/main/java/axel/TaskList.java
#	src/main/java/axel/ToDoTask.java
#	src/main/java/axel/Ui.java

Commit:
Merge branch 'branch-Level-9'

Commit:
Deconflict issues that arose from the merge

Dinoman44 commented on src/main/java/axel/Command.java:
```diff
@@ -0,0 +1,190 @@
+ * Each command manipulates the task list, interacts with the user interface,
+ * and may involve saving changes to storage.
+ */
+public interface Command {
```
Nice naming convention and abstraction! But it may be better to place these classes as separate `.java` files and package them into a `command` package? This way you compartmentalize your code and its easier to debug.

Dinoman44 commented on src/main/java/axel/DeadlineTask.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class DeadlineTask extends Task {
+    /** The deadline for the task. */
+    protected LocalDateTime by;
```
The name kind of makes sense, but maybe you could instead have the name be `deadline` or `endTime`?

Dinoman44 commented on src/main/java/axel/EventTask.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class EventTask extends Task {
+    /** The start time of the event. */
+    protected LocalDateTime from;
```
Same as before, you could probably name this `fromTime` or `startTime`.

Dinoman44 commented on src/main/java/axel/EventTask.java:
```diff
@@ -0,0 +1,58 @@
+    protected LocalDateTime from;
+
+    /** The end time of the event. */
+    protected LocalDateTime to;
```
Same as before, this could be `endTime`.

Dinoman44 commented on src/main/java/axel/Storage.java:
```diff
@@ -0,0 +1,98 @@
+package axel;
+
+import java.io.*;
```
We can't have wildcard imports unfortunately 😭

Dinoman44 commented on src/main/java/axel/TaskList.java:
```diff
@@ -0,0 +1,86 @@
+        return tasks.size();
+    }
+
+    public List<Task> findTasksWithKeyword(String keyword) {
```
CS2030S coming in handy I see. Unfortunately you forgot JavaDoc comment -4 for style 😆

Dinoman44 commented on src/main/java/axel/Axel.java:
```diff
@@ -0,0 +1,80 @@
+package axel;
```
I think its a good idea to have related classes in different packages e.g. a `task` package for task classes, a `command` package for commands, an `axelException` package for your `AxelException`s etc.

Dinoman44 reviewed (commented):
Pretty good job, looking forward to ~~copying~~ seeing your UI! Just some minor naming changes to make.

Commit:
Add a GUI to the chatbot

Commit:
Add assert statements to validate important assumptions across the codebase

Implemented the assert feature in various parts of the code to document and validate key assumptions, improving robustness and making the code easier to maintain. These assertions are used to catch potential logical errors during development and testing.

Parser Class:
* Added assertions to check if the command input is non-null.
* Verified that ToDo descriptions and Deadline commands contain necessary information.

AddCommand Class:
* Ensured the task being added is non-null before proceeding with the execution.

TaskList Class:
* Added bounds-check assertions to verify that the task index is valid when retrieving tasks.

FindCommand Class:
* Added an assertion to verify that the keyword provided for searching tasks is non-null and non-empty.
* Ensured that the list of matching tasks is not null before attempting to print it.

These assertions will help prevent potential issues from propagating through the codebase and are intended to be enabled during testing or debugging by using the -ea flag when running the application.

The use of assertions ensures that key assumptions hold true, leading to better error detection during development. It helps catch invalid states early and improves code quality by making assumptions explicit.

Commit:
Improve code quality

e1121208 committed:
Merge pull request #1 from e1121208/branch-A-Assertions

Add assert statements to validate important assumptions across the codebase

Commit:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/axel/Parser.java

Commit:
Add a 'help' command for in-app user guidance

> Implemented a HelpCommand class that provides a list of available commands and how to use them.
> Updated the Ui class to include a printHelpMessage() method to display the help information.
> Modified the Parser class to handle the "help" command, allowing users to easily access a list of supported commands.

This change improves the usability of the app by providing clear in-app guidance to users.

Commit:
Add a User Guide

Commit:
Revise the `build.gradle` file and add the `MANIFEST.MF` file to ensure that the application runs on different devices

Commit:
Add header comment for a method that was missed out

e1121208 committed:
Update README.md

Commit:
Update `README.md

Commit:
Merge commit '4fc0d8a83592ac990351f9b6aabb5ae2799eaf53'

# Conflicts:
#	README.md

Commit:
Resolve missing GUI title and command bug issues

Commit:
Update READ.md

Commit:
Update README.md
